eos/eos_eospac.cpp

```diff
@@ -0,0 +1,272 @@
+
+#include "eos.hpp"
+#include <eos_Interface.h>
+
+PORTABLE_INLINE_FUNCTION Real TemperatureToSesame(const Real CodeTemp) {
+  return CodeTemp;
+}
+PORTABLE_INLINE_FUNCTION Real TemperatureFromSesame(const Real SesTemp) {
+  return SesTemp;
+}
+PORTABLE_INLINE_FUNCTION Real PressureFromSesame(const Real SesPress) {
+  return 1e10 * SesPress;
+}
+PORTABLE_INLINE_FUNCTION Real PressureToSesame(const Real CodePress) {
+  return 1e-10 * CodePress;
+}
+PORTABLE_INLINE_FUNCTION Real SieToSesame(const Real CodeSie) {
+  return 1e-10 * CodeSie;
+}
+PORTABLE_INLINE_FUNCTION Real SieFromSesame(const Real SesSie) {
+  return 1e10 * SesSie;
+}
+PORTABLE_INLINE_FUNCTION Real CvFromSesame(const Real SesCv) {
+  return 1e10 * SesCv;
+}
+PORTABLE_INLINE_FUNCTION Real BulkModulusFromSesame(const Real SesBmod) {
+  return 1e10 * SesBmod;
+}
+
+namespace singularity {
+
+// How do I make a destructor? How do I free the EOS memory if more than one
+// points to the same table? Does EOSPAC give me multiple (reference-counted)
+// handles to the same table or read in the table multiple times?
+
+EOSPAC::EOSPAC(const int matid) : matid_(matid) {
+  EOS_INTEGER ntables = NT, errorCode = EOS_OK;
+  EOS_INTEGER tableType[NT] = {EOS_Pt_DT, EOS_T_DUt, EOS_Ut_DT, EOS_D_PtT,
+                               EOS_Info};
+  EOS_INTEGER MATID[NT];
+  for (int i = 0; i < NT; i++) {
+    MATID[i] = matid;
+  }
+  eos_CreateTables(&ntables, tableType, MATID, tablehandle, &errorCode);
+  // eos_SetOption(&tablehandle[0],&THE_OPTION,EOS_NullPtr,&errorCode);
+  eos_LoadTables(&ntables, tablehandle, &errorCode);
+  PofRT_table_ = tablehandle[0];
+  TofRE_table_ = tablehandle[1];
+  EofRT_table_ = tablehandle[2];
+  RofPT_table_ = tablehandle[3];
+  EOS_Info_table_ = tablehandle[4];
+
+  // Set reference states
+  EOS_INTEGER error_code = EOS_OK;
+  EOS_INTEGER infoItems[1] = {EOS_Normal_Density};
+  EOS_INTEGER nitems[] = {1};
+  eos_GetTableInfo(&EOS_Info_table_, nitems, infoItems, &rho_ref_, &error_code);
+
+  EOS_REAL R[1] = {rho_ref_};
+  EOS_REAL T[1] = {TemperatureToSesame(temp_ref_)};
+  EOS_REAL E[1], P[1], dx[1], dy[1];
+  EOS_REAL DEDR, DEDT, DPDR, DPDT, DPDE, BMOD;
+  EOS_INTEGER nxypairs = 1;
+  eos_Interpolate(&EofRT_table_, &nxypairs, R, T, E, dx, dy, &errorCode);
+  DEDR = dx[0];
+  DEDT = dy[0];
+  sie_ref_ = SieFromSesame(E[0]);
+  cv_ref_ = CvFromSesame(std::max(DEDT, 0.0));
+
+  eos_Interpolate(&PofRT_table_, &nxypairs, R, T, P, dx, dy, &errorCode);
+  Real PRESS = P[0];
+  press_ref_ = PressureFromSesame(PRESS);
+  DPDR = dx[0];
+  DPDT = dy[0];
+  DPDE = DPDT / (DEDT + EPS);
+  BMOD = rho_ref_ * DPDR + DPDE * (PRESS / (rho_ref_ + EPS) - rho_ref_ * DEDR);
+  bmod_ref_ = BulkModulusFromSesame(std::max(BMOD, 0.0));
+  dpde_ref_ = PressureFromSesame(SieToSesame(DPDE));
+  dvdt_ref_ = dpde_ref_ * cv_ref_ /
+              (rho_ref_ * rho_ref_ * PressureFromSesame(DPDR) + EPS);
+}
+
+PORTABLE_FUNCTION Real EOSPAC::TemperatureFromDensityInternalEnergy(
```

Comment:
Are these actually device callable?

Reply:
No they're not. And cmake will yell at you if you try. I figured I'd leave it though... for... consistency with the other EOS's I suppose.

Reply:
Maybe Cmake preventing it is enough, but it looks a little deceptive. If it's well documented that this isn't GPU capable then it's ok.

Reply:
It's documented. :+1: